File: binary_search_tree.py
class BinarySearchTree:
    def __init__(self, value, depth=1):
        """The __init__ method creates a tree node. The default depth of
        1 equates to the root node.

        Args:
            value (_type_): The value of the node to be added. Can be of
            any type.
            depth (int, optional): How deep the node is in the tree.
            Defaults to 1.
        """

        self.value = value
        self.depth = depth
        self.left = None
        self.right = None

    def insert(self, value):
        """Method to insert a new node in the tree.

        Args:
            value (_type_): The value for the new node.
        """

        # Add the new node to the left
        if value < self.value:
            if self.left is None:
                self.left = BinarySearchTree(value, self.depth + 1)
                print(
                    f"Tree node {value} added to the left of {self.value} at depth "
                    f"{self.depth + 1}"
                )
            else:
                self.left.insert(value)

        # Add the new node to the right
        else:
            if self.right is None:
                self.right = BinarySearchTree(value, self.depth + 1)
                print(
                    f"Tree node {value} added to the right of {self.value} at depth "
                    f"{self.depth + 1}"
                )
            else:
                self.right.insert(value)

    def get_node_by_value(self, value):
        """Method to find a node with given value.

        Args:
            value (_type_): The value of the node to find in the tree.

        Returns:
            _type_: The node that is found
        """
        if self.value == value:
            return self
        elif (self.left is not None) and (value < self.value):
            return self.left.get_node_by_value(value)
        elif (self.right is not None) and (value >= self.value):
            return self.right.get_node_by_value(value)
        else:
            return None

    def find_minimum_value_node(self):
        """Finds the minimum value node in a tree.

        Returns:
            _type_: The node with the minimum value
        """
        current = self.value
        while self.left is not None:
            current = current.left
        return current

    def delete(self, value):
        """Method to remove a node from the tree.

        Args:
            value (_type_): The value of the node to remove.
        """

        # Base case
        if self == None:
            return self

        # If value is less than root value
        if value < self.value:
            self.left = self.left.delete(value)

        # If value is greater than the root value
        elif value > self.value:
            self.right = self.right.delete(value)

        # If the root value is the value to be deleted
        else:
            # If it has no children or one child
            if self.left is None:
                temp = self.right
                self = None
                return temp

            elif self.right is None:
                temp = self.left
                self = None
                return temp

            # A node with two children
            temp = self.right.find_minimum_value_node()
            self.value = temp.value
            self.right.delete(temp)

        return self

    def depth_first_traversal(self):
        """Print the inorder traversal of the Binary Serach Tree."""
        if self.left is not None:
            self.left.depth_first_traversal()
        print(f"Depth={self.depth}, Value={self.value}")

        if self.right is not None:
            self.right.depth_first_traversal()


# For Testing
print("Creating Binary Search Tree rooted at value 15:")
tree = BinarySearchTree(15)

items = [8, 4, 10, 7, 1, 12, 30, 20, 50, 43, 26]
for item in items:
    tree.insert(item)

print("Printing the inorder depth-first traversal:")
tree.depth_first_traversal()
item_to_delete = 7
tree.delete(item_to_delete)
print(f"Printing the inorder depth-first traversal with {item_to_delete} removed:")
tree.depth_first_traversal()

File: javascript/binary_search_tree.js
class BinarySearchTree {

  constructor(value, depth = 1) {
    // Default depth of 1 indicates inital node added to tree is the root
    this.value = value;
    this.depth = depth;
    this.left = null;
    this.right = null;
  }

  insert(value) {
    // "this" refers to the node that called insert().
    // Typically the root node on initial call

    if (value < this.value) {
      // The new node goes to the left of the current node
      if (this.left === null) {
        // Create new node and add as left child of current node
        this.left = new BinarySearchTree(value, this.depth + 1);
        console.log(`Tree node ${value} added to the left of ${this.value} at depth ${this.depth + 1}`);
        return;
      }
      // Call insert() with current node's left child as "this"
      this.left.insert(value);
      return;

    }
    
    // The new node goes to the right of the current node
    if (value > this.value) {

      if (this.right === null) {
        // Create new node and add as right child of current node
        this.right = new BinarySearchTree(value, this.depth + 1);
        console.log(`Tree node ${value} added to the right of ${this.value} at depth ${this.depth + 1}`);
        return;
      }
      
      // Call insert() with current node's right child as "this"
      this.right.insert(value);
      return;
      
    }

    // Values that are already in the tree are ignored

  }

  getNodeByValue(value) {
    // "this" refers to the node that called getNodeByValue().
    // Typically the root node on initial call
    if (this.value === value) {
      return this;
    }
    
    if (this.left !== null && value < this.value) {
      return this.left.getNodeByValue(value);
    }
    
    if (this.right !== null && value > this.value) {
      return this.right.getNodeByValue(value);
    }

    return null;
  }

  findMinValue() {
    // "this" refers to the node that called findMinValue().
    // Typically the root node on initial call
    if (this.left === null) {
      // If there is no left child, this is the minimum node
      return this;
    }
    return this.left.findMinValue();
  }

  findMaxValue() {
    // "this" refers to the node that called findMaxValue().
    // Typically the root node on initial call
    if (this.right === null) {
      // If there is no right child, this is the maximum node
      return this;
    }
    return this.right.findMaxValue();
  }

  delete(value) {
    // "this" refers to the node that called delete().
    // Typically the root node on initial call
    if (this.value === null) {
      return this;
    }

    if (value < this.value) {
      this.left = this.left.delete(value);
    }

    if (value > this.value) {
      this.right = this.right.delete(value);
    }

    if (value === this.value) {

      if (this.left === null) {
        let temp = this.right;
        this.value = null;
        return temp;
      }
      
      if (this.right === null) {
        let temp = this.left;
        this.value = null;
        return temp;
      }
      
      let temp = this.right.findMaxValue();
      this.value = temp.value;
      this.right.delete(temp);

    }
    return this;
  }
  

  depthFirstTraversal() {
    if (this.left !== null) {
      this.left.depthFirstTraversal();
    }
    console.log(`Depth: ${this.depth}, Value: ${this.value}`);
    if (this.right !== null) {
      this.right.depthFirstTraversal();
    }
  }


  printNode() {
    return (`
      Value:       ${this.value}
      Depth:       ${this.depth}
      Left Child:  ${this.left ? this.left.value : null}
      Right Child: ${this.right ? this.right.value : null}
      `);
  }

}

const arr = [8, 4, 10, 7, 1, 30, 20, 50, 43, 26];

const tree = new BinarySearchTree(12);
tree.depthFirstTraversal();

arr.forEach(el => tree.insert(el));
tree.depthFirstTraversal();
console.log(tree.getNodeByValue(20).printNode());
tree.delete(20);
tree.depthFirstTraversal();
console.log('Minimum value: ', tree.findMinValue().printNode());
console.log('Maximum value: ', tree.findMaxValue().printNode());